### DISCUSSION TITLE
One-sided communication

### DISCUSSION THREAD

**Discussion by DaveLanday** [rsmpi/rsmpi#181] [General]
First, I must preface I am a novice user of MPI.

I am using CRDTs (conflict-free replicated data types) in an application and I want to pass data (either a commutative operation in the case of operations based CRDTs or the the entire state of the CRDT in the case of state-based CRDTs) between processes. The beauty of a CRDT is that it's merge semantics are implicit to the data structure and I would like to exploit the properties of CRDTs in an MPI application using One-Sided communication (i.e: RMA). My intuition tells me that I need only One-sided communication since the merge semantics are implicit (i.e: just make a `put` request to add new data to a remote CRDT, and the merge is automatically handled by the owner of the data. Given the "strong eventual consistency" guarantees of CRDTs, and as long as I have configured all host files in a way which creates a network of processes that can reach all other processes given enough `put`s (i.e: a graph with cycles), I only need `put` behavior).

 It says in the README.md that One-sided communication (RMA) is not yet implemented in rsmpi and I was wondering if that was an active area of work / focus for the rsmpi team right now, or if it is low-hanging fruit.
 
 If it is a long way off, are there any potential work arounds? i.e: use rust data types and abstractions specific to sharing data between threads and processes (i.e: Arc, channels, mpsc) to share data to processes on the same node, and rely on two-sided communication methods that rsmpi does already implement for communication between remote nodes?

If I am not thinking about the problem correctly, please let me know, but I would love to hear thoughts. Thanks!

**Comment by jedbrown** [comment]
Good question. Are you only needing `MPI_Put` or also `MPI_Accumulate`? I think an example would be informative.

Are you familiar with MPI one-sided synchronization modes? https://enccs.github.io/intermediate-mpi/one-sided-sync/
One conundrum with one-sided is that active-target one-sided is usually not faster than point-to-point messaging unless you are batching together different operations into the same synchronization epoch (in a way that can't also be batched using point-to-point), but that tends to have rather unsafe semantics. Maybe we can sketch an example and discuss its safety semantics.

**Reply by DaveLanday** [reply]
Oooh! Initially I was thinking `MPI_Put` because it was all I was familiar with (tooling wise). But, `MPI_Accumulate` could also be useful. Consider a CRDT replica `A `on local `process 1` and a `replica B` on remote `process 2` and we want to add an element e to `A` and `B` so that they are in the same state. Would accumulate simultaneously perform the add of `e` to `A` on `process 1` and the `put` to `B` on the remote `process 2`? I can see that being useful. Will accumulate make the one-sided send and call `add e` on `B`, or would I need to call `B.add(e)` (or something like that) once `e` is sent via `MPI_Accumulate` ?

My hope was to use the CRDTs to ease the overhead of global information dissemination that all processes need a view of at any time t (although with eventually consistent views of the global information), so yes I think batching one-sided operations (i.e: when a critical mass of updates to either replica `A` or `B` has been made, or with some probability of passing a message epsilon) would be useful. I am not yet familiar with all useful aspects of MPI tooling yet so any suggestions would be greatly appreciated. Thank you for the links!

**Reply by jedbrown** [reply]
`MPI_Accumulate` can update (atomically per primitive datatype, as far as the MPI application is concerned) values with a built-in operation (`MPI_SUM`, `MPI_LAND`, etc.). https://enccs.github.io/intermediate-mpi/one-sided-routines/#rma-operations
There is no relation between requests to different processes, but no active processing on the target is necessary (except as part of the synchronization protocol).

When everyone has a replica, a standard pattern is to use `MPI_Allreduce` to scalably merge the data from every process. This can be done with a user-defined operation so you can have arbitrarily sophisticated update logic. It can also be nonblocking if using `MPI_Iallreduce`. These operations tend to be much more efficient and scalable than ad-hoc one-sided propagation. If you're pretty sure you still want one-sided, maybe you can describe the data sizes involved, the completion protocol that you envision using, and the characteristic times involved.

**Reply by DaveLanday** [reply]
The size(s) of the data that I want to send/disseminate will not be large. I am hoping to use an operations based CRDT so that I only have to send small buffers between processes (i.e: only the operations for updating the state of a replica is passed and not the entire state of the replica). For example, each process might have it's own work pool and a metric describing the current size or workload of the pool, and I want all processes to have visibility to a all of the other work pool metrics via a map or multi-value register CRDT with Last Writer Wins concurrency semantics. Each process could propagate their updates via a MPI_ibcast and then all receiving broadcasts would `apply` the received updates to their replica using the built in semantics of the CRDT. The other benefit would be the implicit tracking of updates via version vectors (dots I think they are called) for later replaying the (causal) order in which operations were applied etc... Again, this is very much my initial ideas and I am open to feedback, criticism, and directions. As a starting point, I am lookig at this [crate](https://github.com/rust-crdt/rust-crdt/tree/master) for CRDTs in rust

**Reply by jedbrown** [reply]
> all processes to have visibility to a all of the other work pool metrics

This is a non-scalable pattern, but can be done using an `all_to_all`. I don't know how many processes you plan to be working with, but I would suggest doing an initial implementation and then we can look at performance and semantics to see if more hazardous MPI features could help you.

**Reply by DaveLanday** [reply]
Thank you for your feedback. Would it be more scalable to use a gossip-like approach and does MPI provide a gossip protocol type of global message passing? Essentially, if each process has visibility to a "scalable"/manageable number of neighboring processes so that every process is reachable (eventually) from any other process, then eventually all the messages will be passed so that consistency among the replicas is achieved. There could be many processes so a scalable approach would be desirable, but you are right an MVP with some benchmarks is probably first on the to-do list

**Reply by jedbrown** [reply]
That's essentially what collectives do (the portable implementations anyway; vendor implementations sometimes use network topology and special primitives), but you need a completion semantic if you ever need to know that everyone got the result.

**Reply by DaveLanday** [reply]
Ok, thank you for clarifying. I found this [resource](https://www.mpi-forum.org/docs/mpi-1.1/mpi-11-html/node64.html) describing collective communication. Based on what I've (attempted) to describe, is there one that might fit my use case in your opinion? Essentially, I am trying to use CRDTs because they have consistency guarantees whereby as long as a message is sent/received successfully, the states of two replicas communicating with one another will be in the same state. So, my assumption is that the completion semantic (I assume that to mean when collective processes can stop doing work) would just be when all replicas see every message.

**Reply by jedbrown** [reply]
That is from 30 years ago; there are more collectives in the [current standard](https://www.mpi-forum.org/docs/mpi-4.1/mpi41-report/node114.htm#Node114). There is also more tutorial content at [the site](https://enccs.github.io/intermediate-mpi/collective-communication-pt1/) I linked earlier. I think you should encode the update in a user operation. However, I still don't have a clear enough picture of the mutation and completion semantics that your application is going to depend on. I think you should experiment and perhaps come back with a prototype. If there is a simple enough case, you could submit it here as an example (so the use case could be tested and referred to in documentation).

**Reply by DaveLanday** [reply]
You are correct; I will get to work on a hello-world like example using a simple GCounter (grow-only counter). I really appreciate your time and kind responses @jedbrown.

**Reply by DaveLanday** [reply]
Hi @jedbrown, I have an example that compiles but throws an MPI error. I might not be understanding the main concepts of `immediate_ready_send` and `immediate_receive_into`. Essentially, each rank should have it's own `local_counter` which is a `CmRDT::GCounter`. Is this the place for feedback? I didn't open an issue because it is specific to this discussion and thread, but let me know if I should move it on over there.
```rust
#[allow(unused_imports)]
use crdts::{CmRDT, Dot, GCounter};
use mpi::request::WaitGuard;
use mpi::traits::*;
use rand::{thread_rng, Rng};
use std::{thread, time};

// #[derive(Debug, Clone, Equivalence)]
// struct MpiDoti32 {
//     actor: i32,
//     counter: u64,
// }

fn main() {
    #[allow(unused_variables)]
    // number of fake jobs to process (i.e: count up to this number)
    const NJOBS: u16 = 2000;

    // Represents the worst time job can take to finish
    let dur = time::Duration::from_millis(25);

    // Represents the probability of performing a large job:
    const P: f64 = 0.015; // 15% chance of a long running job (i.e: 2.5 seconds)
    let mut rng = thread_rng();

    // Init MPI
    let universe = mpi::initialize().unwrap();
    let world = universe.world();

    // Get the rank and size of the communicator
    let size = world.size();
    let rank = world.rank();

    // Define the processor that sends to this rank and  that receives from this rank:
    let prev_rank = (rank - 1) % size;
    let next_rank = (rank + 1) % size;

    // Init local replica of the count
    let mut local_counter = GCounter::new();

    // We need a mutable buffer to send operations from the local process
    // let mut local_op = MpiDoti32 {
    //     actor: rank.clone(),
    //     counter: 0,
    // };

    // For getting the most up-to-date count of the remote Dot
    let mut remote_counter = vec![0u64];

    for _n in 0..NJOBS {
        // simulate a long or short process between incrementing the count:
        if rng.gen_bool(P) {
            thread::sleep(dur);
            println!("Running a long job on processor {}", rank);
        }

        // Declare intent to increment by creating an `Op`. Use rank as the uid of the replica
        let inc_op = local_counter.inc(&rank);

        // Apply the operation to our local counter first and then send the `Op` off to others
        local_counter.apply(inc_op.clone()); // Apply the increment:

        // Format the local operation for sending:
        // local_op.counter = inc_op.counter.clone();

        // Simple point-to-point messaging in a ring configuration:
        mpi::request::scope(|scope| {
            // Receive the operation to increment the counter from the previous rank:
            let _rreq = WaitGuard::from(
                world
                    .process_at_rank(prev_rank)
                    .immediate_receive_into(scope, &mut remote_counter),
            );

            // Send the change that was applied by the local counter:
            let _sreq = WaitGuard::from(
                world
                    .process_at_rank(next_rank)
                    .immediate_ready_send(scope, &inc_op.counter),
            );
        });
        // Apply the change to the local counter:
        local_counter.inc_many(&prev_rank, remote_counter[0]);
    }
    println!(
        "Processor {} says the count = {:?}",
        rank,
        local_counter.read()
    );
}
```
The following is thrown when I call `mpirun -np 4 target/debug/<my_program>`:
```shell
❯ mpirun -np 4 target/debug/MPI_Plotter
thread 'main' panicked at /<path>/.cargo/registry/src/index.crates.io-6f17d22bba15001f/mpi-0.7.0/src/topology/mod.rs:563:9:
assertion failed: 0 <= r && r < self.target_size()
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[<Computer>:00000] *** An error occurred in MPI_Wait
[<Computer>:00000] *** reported by process [2931163137,3]
[<Computer>:00000] *** on communicator MPI_COMM_WORLD
[<Computer>:00000] *** MPI_ERR_TRUNCATE: message truncated
[<Computer>:00000] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[<Computer>:00000] ***    and MPI will try to terminate your MPI job as well)
```

**Reply by DaveLanday** [reply]
Ok, so I made a mistake defining the `prev_rank`. It should be `prev_rank = (rank + size - 1) % size;`. I also got a suggestion from one of the maintainers of crdts.rs crate and they suggested that it might be easier to serialize the `op` to `Vec<u8>` using bincode so I don't have to implement `Equivalence` for `Dot` in their library. Here is my edited code and output:
```rust
#[allow(unused_imports)]
use crdts::{CmRDT, Dot, GCounter};
use mpi::request::WaitGuard;
use mpi::traits::*;
use rand::{thread_rng, Rng};
use std::{thread, time};
use bincode;

// #[derive(Debug, Clone, Equivalence)]
// struct MpiDoti32 {
//     actor: i32,
//     counter: u64,
// }

fn main() {
    #[allow(unused_variables)]
    // number of fake jobs to process (i.e: count up to this number)
    const NJOBS: u16 = 2000;

    // Represents the worst time job can take to finish
    let dur = time::Duration::from_millis(25);

    // Represents the probability of performing a large job:
    const P: f64 = 0.015; // 15% chance of a long running job (i.e: 2.5 seconds)
    let mut rng = thread_rng();

    // Init MPI
    let universe = mpi::initialize().unwrap();
    let world = universe.world();

    // Get the rank and size of the communicator
    let size = world.size();
    let rank = world.rank();

    // Define the processor that sends to this rank and  that receives from this rank:
    let prev_rank = (rank + size - 1) % size;
    let next_rank = (rank + 1) % size;

    // Init local replica of the count
    let mut local_counter = GCounter::new();

    println!(
        "Prev rank: {}, This rank: {}, Next rank: {}",
        prev_rank, rank, next_rank
    );
    for _n in 0..NJOBS {
        // simulate a long or short process between incrementing the count:
        if rng.gen_bool(P) {
            thread::sleep(dur);
            println!("Running a long job on processor {}", rank);
        }

        // Declare intent to increment by creating an `Op`. Use rank as the uid of the replica
        let inc_op = local_counter.inc(rank);

        // Apply the operation to our local counter first and then send the `Op` off to others
        local_counter.apply(inc_op.clone()); // Apply the increment:

        // Serialize the local operation for sending:
        let send_op: Vec<u8> = bincode::serialize(&inc_op).unwrap();

        // For getting the most up-to-date count of the remote Dot
        let mut remote_op: Vec<u8> = Vec::new();

        // Simple point-to-point messaging in a ring configuration:
        mpi::request::scope(|scope| {
            // Receive the operation to increment the counter from the previous rank:
            let _rreq = WaitGuard::from(
                world
                    .process_at_rank(prev_rank)
                    .immediate_receive_into(scope, &mut remote_op),
            );

            // Send the change that was applied by the local counter:
            let _sreq = WaitGuard::from(
                world
                    .process_at_rank(next_rank)
                    .immediate_ready_send(scope, &send_op),
            );
        });

        // Deserialize the remote_op:
        let recv_op: Dot<mpi::Rank> = bincode::deserialize(&remote_op).unwrap();

        // Apply the change to the local counter:
        local_counter.apply(recv_op);
    }
    println!(
        "Processor {} says the count = {:?}",
        rank,
        local_counter.read()
    );
}
```
Now the print statement directly before the `for-loop` executes on all the processes, but I still get the MPI_WAIT error

**Reply by DaveLanday** [reply]
I ended up commenting on an open issue: https://github.com/rsmpi/rsmpi/issues/144#issuecomment-2052429220
